### PR TITLE
Node defs validation

### PR DESCRIPTION
**Pull request by SteRiccio**
## Description

- #### Added SurveyNodeDefIndex
Improved the speed of survey with node defs loading and dependency graph generation by adding an index (node defs uuids by parent uuid). The function Survey.getNodeDefChildren defined in SurveyNodeDefs is called quite frequenty during the evaluation of expressions. Without an index on the node definition, an iteration among all the node defs is necessary, comparing their uuids.
The node defs index can be initialized every time a survey and its node defs are loaded and it is updated on overy node def add or remove.

## Types of changes

- [ ] Docs change
- [x] Refactoring
- [ ] Dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Do you consider this PR needs further testing?

- [ ] No
- [x] Yes

Unit test on SurveyNodeDefIndex module should be defined.

### CODE REVIEW

**Comment by minotogna** [comment]
Task linked: [CU-jda9ww NodeDefs validation](https://app.clickup.com/t/jda9ww)

**Review by sorja — COMMENTED** [review]
👍

File: core/survey/_survey/surveyNodeDefsIndex.js [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+export const assocNodeDefsIndex = ({ survey, nodeDefsIndex }) => ({ ...survey, nodeDefsIndex })
+
+// ==== UPDATE
+
+export const addNodeDefToIndex = ({ nodeDefsIndex, nodeDef }) => {
+  const nodeDefsIndexUpdated = { ...nodeDefsIndex }
+  // add node def uuid to parent node def children references
+  const { parentUuid } = nodeDef
+  if (parentUuid && !nodeDefsIndexUpdated.childDefUuidsByParentUuid[parentUuid].includes(nodeDef.uuid)) {
```
Comment (sorja) [inline core/survey/_survey/surveyNodeDefsIndex.js:61]: maybe null check here: `if (parentUuid && !nodeDefsIndexUpdated?.childDefUuidsByParentUuid?.[parentUuid]?.includes(nodeDef.uuid)) {`

Comment (SteRiccio) [reply core/survey/_survey/surveyNodeDefsIndex.js:61]: yes thanks, I've already changed it

File: core/survey/_survey/surveyDependencies.js [hunk 2]
```diff
@@ -16,10 +16,11 @@ const keys = {
 
 const _getDeps = (type, nodeDefUuid) => R.pathOr([], [type, nodeDefUuid])
 
-const _addDep = (type, nodeDefUuid, nodeDefDepUuid) => (graph) =>
-  R.pipe(_getDeps(type, nodeDefUuid), R.append(nodeDefDepUuid), (dep) =>
-    ObjectUtils.setInPath([type, nodeDefUuid], dep)(graph)
-  )(graph)
+const _addDep = (type, nodeDefUuid, nodeDefDepUuid) => (graph) => {
+  const deps = _getDeps(type, nodeDefUuid)(graph)
+  const depsUpdated = [...deps, nodeDefDepUuid]
```
Comment (ramirobg94) [inline core/survey/_survey/surveyDependencies.js:21]: much easier to read:

take a look of:
https://stackoverflow.com/questions/48865710/spread-operator-vs-array-concat

File: server/modules/nodeDef/service/nodeDefService.js [hunk 3]
```diff
@@ -1,6 +1,134 @@
...
+
+export const updateNodeDefProps = async (
+  user,
+  surveyId,
+  cycle,
+  nodeDefUuid,
+  parentUuid,
+  props,
+  propsAdvanced = {},
+  system = false,
```
Comment (ramirobg94) [inline server/modules/nodeDef/service/nodeDefService.js:92]: do we have to add object here?

File: server/modules/nodeDef/service/nodeDefService.js [hunk 4]
```diff
@@ -1,6 +1,134 @@
...
+      advanced: true,
+      validate: false,
+    },
+    t
+  )
+  // build dependency graph if empty
+  return Survey.hasDependencyGraph(surveyDb) ? surveyDb : Survey.buildAndAssocDependencyGraph(surveyDb)
+}
+
+const afterNodeDefUpdate = async ({ survey, nodeDef, nodeDefsDependent = [], nodeDefsUpdated }, t) => {
```
Comment (ramirobg94) [inline server/modules/nodeDef/service/nodeDefService.js:25]: do the t should be "client = db"

File: server/modules/nodeDef/service/nodeDefService.js [hunk 5]
```diff
@@ -1,6 +1,134 @@
...
+import * as NodeDef from '@core/survey/nodeDef'
+import * as SurveyValidator from '@core/survey/surveyValidator'
+import * as Validation from '@core/validation/validation'
+
+import { db } from '@server/db/db'
+import * as SurveyManager from '@server/modules/survey/manager/surveyManager'
 import * as NodeDefManager from '../manager/nodeDefManager'
 
-export const { insertNodeDef, updateNodeDefProps, markNodeDefDeleted } = NodeDefManager
+const fetchSurvey = async ({ surveyId, cycle }, t) => {
```
Comment (ramirobg94) [inline server/modules/nodeDef/service/nodeDefService.js:10]: do the t should be client = db

File: server/modules/nodeDef/service/nodeDefService.js (hunk 3 above)
Comment (SteRiccio) [reply server/modules/nodeDef/service/nodeDefService.js:92]: yes, it's better!

File: server/modules/nodeDef/service/nodeDefService.js (hunk 4 above)
Comment (SteRiccio) [reply server/modules/nodeDef/service/nodeDefService.js:25]: it is an internal function, I didn't want to change too much the code, but yes, you are right

File: server/modules/nodeDef/service/nodeDefService.js (hunk 5 above)
Comment (SteRiccio) [reply server/modules/nodeDef/service/nodeDefService.js:10]: same as above ;)

**Review by ramirobg94 — APPROVED** [review]

**Pull request merged**